File: concurrent-sum-java/src/main/java/org/example/SumCalc.java
package org.example;

public class SumCalc {
    private final long[] arr;
    private final int chunkSize;

    public SumCalc(long[] arr, int numSplits) {
        this.arr = arr;
        this.chunkSize = (int) Math.ceil((double) arr.length / numSplits);
    }

    public long subArrSum(int start, int end) {
        long sum = 0;
        for (int i = start; i < end; i++) {
            sum += arr[i];
        }
        return sum;
    }

    public int[] getChunkBounds(int index) {
        int start = index * chunkSize;
        int end = Math.min(start + chunkSize, arr.length);
        return new int[]{start, end};
    }

}

File: concurrent-sum-kotlin/src/main/kotlin/SumCalc.kt
package org.example

class SumCalc(val arr: LongArray, numSplits: Int) {
    private val chunkSize: Int = kotlin.math.ceil(arr.size.toDouble() / numSplits).toInt()

    fun subArrSum(start: Int, end: Int): Long {
        var sum = 0L
        for (i in start until end) {
            sum += arr[i]
        }
        return sum
    }

    fun getChunkBounds(index: Int): IntArray {
        val start = index * chunkSize
        val end = kotlin.math.min(start + chunkSize, arr.size)
        return intArrayOf(start, end)
    }
}
